### PR TITLE
Dev avinash tips

### PR DESCRIPTION
**Pull request by avinashpatel57**

### CODE REVIEW

File: backend/stylists/app/Http/Mapper/TipMapper.php [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+        foreach ($this->dropdown_fields as $dropdown_field) {
+            $tip->$dropdown_field = isset($request->$dropdown_field) && $request->$dropdown_field != '' ? $request->$dropdown_field : '';;
+        }
+        return $tip;
+    }
+
+    public function saveEntities($tip_id, $products, $looks)
+    {
+        $entity_product_ids = $products ? explode(',', $products) : [];
```
Comment (techstunts) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: $new_product_ids

File: backend/stylists/app/Http/Mapper/TipMapper.php [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        $entity_look_ids = $looks ? explode(',', $looks) : [];
+
+        $tip_entities = $this->getExistingEntities($tip_id);
+
+        $existing_look_entities = [];
```
Comment (techstunts) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: $existing_look_ids

File: backend/stylists/app/Http/Mapper/TipMapper.php [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+                $status = false;
+                $message = $e->getMessage();
+                return array(
+                    'status' => $status,
+                    'message' => $message,
+                );
+            }
+        }
+
+        $raw_query = '';
```
Comment (techstunts) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: $delete_entities_query

File: backend/stylists/app/Http/Mapper/TipMapper.php [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+        $products_to_delete = array_diff($existing_product_entities, $entity_product_ids);
+        $looks_to_delete = array_diff($existing_look_entities, $entity_look_ids);
+        $products_to_add = array_diff($entity_product_ids, $existing_product_entities);
+        $looks_to_add = array_diff($entity_look_ids, $existing_look_entities);
+
+        $entity_array = [];
+        $index = 0;
+        foreach ($products_to_add as $entity_product_id) {
+            $entity_array[$index++] = array(
```
Comment (techstunts) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: $insert_entities

File: backend/stylists/app/Http/Mapper/TipMapper.php [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+        $raw_query = '';
+        foreach ($products_to_delete as $item) {
+            $raw_query = $raw_query . " OR (tip_id = '{$tip_id}' AND entity_type_id = '{$entity_type_product}' AND entity_id = '{$item}')";
+        }
+        foreach ($looks_to_delete as $item) {
+            $raw_query = $raw_query . " OR (tip_id = '{$tip_id}' AND entity_type_id = '{$entity_type_look}' AND entity_id = '{$item}')";
+        }
+
+        if (!empty($raw_query)) {
```
Comment (techstunts) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: Keep both insert and delete statements together in a single try catch. Also use a transaction to execute these queries.

File: backend/stylists/app/Http/Controllers/TipController.php [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        if ($validator->fails()) {
+
+            return Redirect::back()
+                ->withErrors($validator)
+                ->withInput($request->all());
+        }
+
+        $tip = new Tip();
+
+        $tip = $tipMapperObj->setObjectProperties($tip, $request);
```
Comment (techstunts) [inline backend/stylists/app/Http/Controllers/TipController.php]: Move this code into TipMapper

File: backend/stylists/app/Http/Controllers/TipController.php [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        $validator = $tipMapperObj->inputValidator($request);
+
+        if ($validator->fails()) {
+            return redirect('tip/edit/' . $this->resource_id)
+                ->withErrors($validator)
+                ->withInput($request->all());
+        }
+
+
+        $tip = Tip::find($this->resource_id);
```
Comment (techstunts) [inline backend/stylists/app/Http/Controllers/TipController.php:201]: Move this code to TipMapper

File: backend/stylists/public/js/datatable.js [hunk 8]
```diff
@@ -268,20 +271,90 @@ $(document).ready(function () {
...
+
+        if (checked_items.length <= 0) {
+            alert('Please select at least one item');
+            return false;
+        }
+
+        var cross_mark = '<span class="pull-right cross_mark"><a href="#"><i class="material-icons" style="font-size: 13px;">close</i></a></span>';
+
+        if (entity_type_id == EntityType.LOOK) {
+            checked_items.parent().parent().each(function () {
```
Comment (techstunts) [inline backend/stylists/public/js/datatable.js]: Try siblings([selector]), if doesnt works, try parents([selector])

File: backend/stylists/public/js/datatable.js [hunk 9]
```diff
@@ -268,20 +271,90 @@ $(document).ready(function () {
...
+                var inputChkBox = $(this).children('.name').children('input');
+
+                $(this).children('.name').prepend(cross_mark);
+                $(this).children('.name').children('span').on('click', deleteItem);
+                $(this).attr('value', inputChkBox.val());
+                inputChkBox.remove();
+                $(this).appendTo($("#look_ids").parent().children('.content')[0]);
```
Comment (techstunts) [inline backend/stylists/public/js/datatable.js]: Try siblings([selector])

File: backend/stylists/public/js/datatable.js [hunk 10]
```diff
@@ -268,20 +271,90 @@ $(document).ready(function () {
...
+            $(".mobile-app-send .btn").removeClass('active');
+            $(".mobile-app-send .btn").addClass('disabled');
+        }
+        alert('Items added in list');
+    });
+
+    $(".info").find('input:submit').on('click', function(){
+        var look_ids = [];
+
+        $("#look_ids").parent().children('.content').find('.items').each(function(){
```
Comment (techstunts) [inline backend/stylists/public/js/datatable.js]: Use this to select look_ids under current form

File: backend/stylists/app/Http/Mapper/TipMapper.php (hunk 1 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: done

File: backend/stylists/app/Http/Mapper/TipMapper.php (hunk 2 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: done

File: backend/stylists/app/Http/Mapper/TipMapper.php (hunk 3 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: done

File: backend/stylists/app/Http/Mapper/TipMapper.php (hunk 4 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: done

File: backend/stylists/app/Http/Mapper/TipMapper.php (hunk 5 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Mapper/TipMapper.php]: done

File: backend/stylists/app/Http/Controllers/TipController.php (hunk 6 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Controllers/TipController.php]: done

File: backend/stylists/public/js/datatable.js (hunk 10 above)
Comment (avinashpatel57) [inline backend/stylists/public/js/datatable.js]: done

File: backend/stylists/public/js/datatable.js (hunk 9 above)
Comment (avinashpatel57) [inline backend/stylists/public/js/datatable.js]: done

File: backend/stylists/app/Http/Controllers/TipController.php (hunk 7 above)
Comment (avinashpatel57) [inline backend/stylists/app/Http/Controllers/TipController.php:201]: done

File: backend/stylists/public/js/datatable.js (hunk 8 above)
Comment (avinashpatel57) [inline backend/stylists/public/js/datatable.js]: done

**Pull request merged**